docs/architecture/adr-008-restructure-ibc-crate.md

```diff
@@ -0,0 +1,122 @@
+
+As of [this pull request](https://github.com/cosmos/ibc-rs/pull/954), we've separated our mock testing kit into the standalone **`ibc-testkit`** library. This decoupling from the main **`ibc`** crate sets the stage for the objectives of this ADR.
+
+The primary goals here are twofold: firstly, to reduce interdependence within the codebase among various components such as ibc core, clients, and applications, and secondly, to improve the overall usability of the `ibc-rs` implementation. The overarching aim is to empower users to selectively import specific IBC layers, mitigating potential conflicts related to dependencies or features that may arise in the context of a monolithic library and letting `ibc-rs` be used in the following scenarios:
+
+1. **Selective Module Import**
+    - Users cannot import only the necessary components/modules for their projects. For instance, importing only the **`ics07_tendermint`** implementation is impractical.
+2. **Selective Types Import**
+    - Relayers, like Hermes, or any off-chain consumers cannot import their desired layer of implementation like ibc types without pulling in unnecessary dependencies into their project.
+3. **Smoother IBC Core Integration with Hosts**
+    - Integrating ibc core with host chains without introducing light client or app dependencies is currently not straightforward, impeding smooth integration.
+4. **Easier Development of CosmWasm Contracts**
+    - For developing a CosmWasm tendermint light client, we ideally should only be dependent on implementation under the **`ics07_tendermint`** and also be importing relevant parts from the **`ibc-core-client`** layer without pulling in all the ibc codebase and dependencies.
+
+This ADR aims to enhance both the usability and practicality of `ibc-rs` by restructuring the codebase and organizing it under multiple sub-libraries, as stated in the [decision](#decision) section. This will make different parts of `ibc-rs` accessible to users, positioning it as a more comprehensive, one-stop solution catering to diverse users groups, whether for on-chain or off-chain use cases.
+
+## Decision
+
+For the library organization, the first stage of separation is to split off the codebase of each IBC applications, clients, and core implementation, decoupling them from each other. The top-level libraries and the naming schema would look like as follow:
+
+```markdown
+.
+├── ibc -> Primarily re-exports sub-libraries
+├── ibc-core 
+│   ├── ibc-core-client (contains the implementation + Re-exports types)
+│   ├── ibc-core-connection
+│   ├── ibc-core-channel
+│   ├── ibc-core-commitment
+│   └── ibc-core-host
+│       └── .
+├── ibc-clients
+│   ├── ibc-client-tendermint
+│   ├── ibc-client-tendermint-cw
```

Comment:
Should this be `ibc-client-cosmwasm`?

Reply:
I was in favor of setting up a whole new crate for Wasm clients at first. Though, in practice I am seeing that's just extra crate we don't need. We can keep it simple and throw all the light clients implementation under the `ibc-clients`. And to tell apart the CosmWasm implementation of a specific light client, let's just append a `-cw` at the end. Therefore for `ibc-client-tendermint` the CW implementation would be maintained under `ibc-client-tendermint-cw`